builtin/update-index.c

```diff
@@ -411,6 +413,12 @@ static int add_cacheinfo(unsigned int mode, const struct object_id *oid,
 	if (!verify_path(path, mode))
 		return error("Invalid path '%s'", path);
 
+	if (S_ISSPARSEDIR(mode))
+		return error("Cannot add directory '%s' as cache entry", path);
```

Comment:
Interesting that we rely on the input mode. This will fail any time `mode` is a directory. Is there not already a protection for that?

also, messages in `error()`, `warning()`, or `die()` calls should start with a lowercase letter.

Reply:
I double-checked, and it turns out there weren't - it was entirely possible to add a sparse directory and have it cause bugs in the index & cache tree (_especially_ in non-sparse indexes). I moved that fix (and a test for it in `t2107-update-index-basic`) into a dedicated commit at the beginning of this series and updated the `update-index --cacheinfo` test so that, without this change, it would have "succeeded" in inserting the directory entry into the index (in other words, I wasn't properly testing it without the `--add` flag).